### PR TITLE
[WIP] Create a distance sensor driver base class

### PR DESCRIPTION
**Pull request by mcsauder**
**Describe problem solved by the proposed pull request**
This PR creates a DistanceSensor base class, and refactors the following drivers to inherit from the base class
- [x] cm8jl65
- [x] hc_sr04
- [x] LeddarOne
- [ ] LidarLite (ll40ls)
- [ ] MapppyDot
- [ ] mb12xx
- [ ] pga460
- [ ] sf0x
- [ ] sf1xx
- [ ] srf02
- [ ] teraranger
- [ ] tfmini
- [x] uLanding
- [ ] vl53lxx

**Test data / coverage**
Test data will be presented...

**Describe problem solved by the proposed pull request**
The distance sensor drivers are largely copy/paste from previous implementations.  This PR establishes a distance sensor driver base class to allow an inheritance, better distance design pattern and architecture, better boilerplate code for future implementations, and less overall code duplication.

**Describe possible alternatives**
The drivers certainly can be left as-is.

**Additional context**
See #11853. #11891,  #9279, #11853, #11857, #1858, #11859. etc.

### CODE REVIEW

**Comment by dagar** [comment]
@davids5 do you know where we can purchase this sensor?

**Comment by mcsauder** [comment]
@dagar , and @davids5 , the closest product number I can find to the CM8JL65 on Lanbao's site are the PTK series: http://www.lanbaosensor.com/index.php?m=content&c=index&a=prolists&catid=96.  I have not been able to locate a source or a product page for the CM8JL65.

**Comment by LorenzMeier** [comment]
It is a custom build that will be made available in the next couple of weeks. It is much smaller compared to their previous offerings.

**Comment by cmic0** [comment]
@mcsauder at the moment all the vehicles we have that mount this distance sensor are used for testing other functionalities. Will try to allocate some time this week.

**Comment by davids5** [comment]
> @davids5 do you know where we can purchase this sensor?

@dagar Sorry No I do not

**Comment by bys1123** [comment]
This looks very similar with TFminiPlus

**Comment by cmic0** [comment]
@mcsauder the distance sensor is now available on seeed
https://www.seeedstudio.com/PSK-CM8JL65-CC5-ToF-Infrared-Distance-Measuring-Sensor-p-4028.html

I'll give a shot to your branch later today

**Comment by mcsauder** [comment]
Thanks @cmic0 , I have a sensor pre-ordered now, so in about a month I will be able to test this work.  If you have spare cycles I am grateful for any feedback you have, but feel free to wait and I will update this PR with test data when my sensor arrives.  Thanks for the link!

File: src/lib/drivers/distance_sensor/DistanceSensor.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+{
+	perf_begin(_sample_perf);
+
+	if (!_collect_phase) {
+		return measure();
+	}
+
+	distance_sensor_s report;
+	report.current_distance = get_distance();
+	report.id               = get_sensor_id();
```
Comment (dagar) [inline src/lib/drivers/distance_sensor/DistanceSensor.cpp]: Most of these fields don't change every iteration.

File: src/lib/drivers/distance_sensor/DistanceSensor.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+int
+DistanceSensor::measure()
+{
+	return PX4_OK;
+}
+
+int
+DistanceSensor::open_serial_port(const speed_t speed)
```
Comment (dagar) [inline src/lib/drivers/distance_sensor/DistanceSensor.cpp]: I think handling a serial port is an orthogonal piece.

File: src/lib/drivers/distance_sensor/DistanceSensor.h [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+#include <uORB/topics/distance_sensor.h>
+#include <uORB/uORB.h>
+
+
+using namespace time_literals;
+
+/* Configuration Constants */
+#define DEFAULT_BAUD_RATE         B115200      // Baudrate 115200.
+#define DEFAULT_MEASURE_INTERVAL  100_ms       // 100ms default sensor measurement interval.
+#define DEFAULT_SERIAL_PORT       "/dev/ttyS3" // Default UART port.
```
Comment (dagar) [inline src/lib/drivers/distance_sensor/DistanceSensor.h]: The entire concept of default port probably needs to go away  from the driver perspective and be set per board.

File: src/lib/drivers/distance_sensor/DistanceSensor.cpp (hunk 1 above)
Comment (mcsauder) [reply src/lib/drivers/distance_sensor/DistanceSensor.cpp]: Agreed. I'll rework this.  Thanks for looking this over!

File: src/lib/drivers/distance_sensor/DistanceSensor.cpp (hunk 2 above)
Comment (mcsauder) [reply src/lib/drivers/distance_sensor/DistanceSensor.cpp]: Agreed.  I'll break this out into a separate implementation.

File: src/lib/drivers/distance_sensor/DistanceSensor.h (hunk 3 above)
Comment (mcsauder) [reply src/lib/drivers/distance_sensor/DistanceSensor.h]: OK. For now I'll pull this out into a separate `SerialPort` class along with the `open_serial_port()` implementation.

**Comment by mcsauder** [comment]
I think I'll close this PR and work on adding functionality and porting distance sensor drivers over to inherit from the PX4RangeFinder class.

**Pull request closed without merging**